content/blog/scheduling-in-react/index.md

```diff
@@ -0,0 +1,243 @@
+
+The application works like this: The more you type into the input below, the more detailed the charts below will get. Since both of the updates (the input element and the chart) run at the same time, the browser has to do so much computation that it will drop some frames. This leads to noticeable delays and a bad user experience:
+
+<video src="/blog/scheduling-in-react/sync-mode.mp4" muted="true" autoplay muted playsinline loop></video>
+
+However, a version that prioritizes updating the input with new keystrokes will appear to the end user as though it’s running a lot faster. This is because users receive immediate feedback even though the same computation time is required:
+
+<video src="/blog/scheduling-in-react/concurrent-mode.mp4" muted="true" autoplay muted playsinline loop></video>
+
+Unfortunately, current user interface architectures make it non-trivial to implement prioritization. One way to work around this problem is by [debouncing](https://davidwalsh.name/javascript-debounce-function) the chart update. The problem with this approach is that the charts still render synchronously when the debounced callback fires, which will again cause the user interface to be unresponsive for some time. We can do better!
+
+## Browser Event Loop
+
+Before we learn more about how proper prioritizing of updates can be achieved, let’s dig deeper and understand why the browser has issues with these kind of user interactions.
+
+JavaScript code is executed in one thread, meaning that only one line of JavaScript can be run at any given time. The same thread is also responsible for other document lifecycles, like layout and paint.[^1] This means that whenever JavaScript code runs, the browser is blocked from doing anything else.
+
+To keep the user interface responsive, we only have a very short timeframe before we need to be able to receive the next input events. At the Chrome Dev Summit 2018, Shubhie Panicker and Jason Miller gave a talk, [A Quest to Guarantee Responsiveness](https://developer.chrome.com/devsummit/schedule/scheduling-on-off-main-thread). During the talk, they presented the browser’s run loop visualization, in which we can see that we only have 16ms (on a typical 60Hz screen) before the next frame is drawn and the next events need to be processed:
+
+![The browser event loop starts by running input handlers. Then it runs animation frame callbacks, and it ends with document lifecycles (style, layout, paint). All of this should complete within one frame, which is approximately 16ms on a 60Hz display.](event-loop-browser.png)
+
+Most JavaScript frameworks (including the current version of React) will run updates synchronously. We can think of this as a function `render()` that will only return once the DOM is updated. During this time, the main thread is blocked.
```

Comment:
What is "this?" The frameworks running updates synchronously? How would we describe that in the following sentence? A functionality? A feature? Something else?

"We can think of this as a function `render()` that will only return once the DOM was updated. During this time, the main thread is blocked."

Reply:
I'm leaning towards "behavior" or "implementation" although I think the latter might be more correct but it kinda feels wrong and I don't know why.

Reply:
I like behavior better so I'll put that in, as you're right, implementation doesn't necessarily feel correct.